libraries/chain/finality/qc.cpp

```diff
@@ -0,0 +1,507 @@
+        ("c", connection_id)("bn", block_num)("sv", strong)("s", s)("pre", pre_state)("state", post_state)("q", is_quorum_met(post_state)));
+   return s;
+}
+
+// called by get_best_qc which acquires a mutex
+qc_sig_t aggregating_qc_sig_t::extract_qc_sig_from_aggregating() const {
+   qc_sig_t qc_sig;
+
+   if( aggregating_state == state_t::strong ) {
+      qc_sig.strong_votes = strong_votes.bitset;
+      qc_sig.sig          = strong_votes.sig;
+   } else if (is_quorum_met_no_lock()) {
+      qc_sig.strong_votes = strong_votes.bitset;
+      qc_sig.weak_votes   = weak_votes.bitset;
+      qc_sig.sig          = strong_votes.sig;
+      qc_sig.sig.aggregate(weak_votes.sig);
+   } else
+      assert(0); // this should be called only when we have an aggregating_qc_sig_t with a quorum
+
+   return qc_sig;
+}
+
+std::optional<qc_sig_t> aggregating_qc_sig_t::get_best_qc() const {
+   std::lock_guard g(*_mtx);
+   // if this does not have a valid QC, consider received_qc_sig only
+   if( !is_quorum_met_no_lock() ) {
+      if( received_qc_sig ) {
+         return std::optional{*received_qc_sig};
+      }
+      return {};
+   }
+
+   qc_sig_t qc_sig_from_agg = extract_qc_sig_from_aggregating();
+
+   // if received_qc_sig does not have value, consider qc_sig_from_agg only
+   if( !received_qc_sig ) {
+      return std::optional{std::move(qc_sig_from_agg)};
+   }
+
+   // Both received_qc_sig and qc_sig_from_agg have value. Compare them and select a better one.
+   // Strong beats weak. Tie-break by received_qc_sig, strong beats weak
+   if (received_qc_sig->is_strong() || qc_sig_from_agg.is_weak()) {
+      return std::optional{qc_sig_t{ *received_qc_sig }};
+   }
+   return std::optional{qc_sig_t{ std::move(qc_sig_from_agg) }};
+}
+
+bool aggregating_qc_sig_t::set_received_qc_sig(const qc_sig_t& qc) {
+   std::lock_guard g(*_mtx);
+   if (!received_qc_sig || (received_qc_sig->is_weak() && qc.is_strong())) {
+      received_qc_sig = qc;
+      return true;
+   }
+   return false;
+}
+
+bool aggregating_qc_sig_t::received_qc_sig_is_strong() const {
+   std::lock_guard g(*_mtx);
+   return received_qc_sig && received_qc_sig->is_strong();
+}
+
+bool aggregating_qc_sig_t::is_quorum_met_no_lock() const {
+   return is_quorum_met(aggregating_state);
+}
+
+std::optional<qc_t> aggregating_qc_t::get_best_qc(block_num_type block_num) const {
+   std::optional<qc_sig_t> active_best_qc = active_policy_sig.get_best_qc();
+   if (!active_best_qc) // active is always required
+      return {};
+
+   if (pending_policy_sig) {
+      std::optional<qc_sig_t> pending_best_qc = pending_policy_sig->get_best_qc();
+      if (pending_best_qc)
+         return std::optional<qc_t>{qc_t{block_num, std::move(*active_best_qc), std::move(pending_best_qc)}};
+      return {}; // no quorum on pending_policy_sig so no qc for this block
```

Comment:
Because that is not a valid QC if there is a pending finalizer policy, we must reach quorum on pending as well as on the active finalizer policy otherwise we do not have a valid QC.